rpc tests leak clients

The following test appears to have failed:

[#3910](https://circleci.com/gh/cockroachdb/cockroach/3910):

```
=== RUN TestUnretryableError
I0623 02:18:40.893453     337 client.go:121] client 127.0.0.1:51115 connected
--- PASS: TestUnretryableError (0.32s)
=== RUN TestClientNotReady
E0623 02:18:41.207007     337 client.go:118] client 127.0.0.1:58314 failed to connect: client.go:303: client is closed
--- FAIL: TestClientNotReady (0.41s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080118e0, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x63
		net.(*pollDesc).WaitRead(0xc2080118e0, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x51
		net.(*netFD).Read(0xc208011880, 0xc2083aa800, 0x400, 0x400, 0x0, 0x7f9b4ee1fd80, 0xc2084aa018)
			/usr/src/go/src/net/fd_unix.go:242 +0x4b3
		net.(*conn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0x4403fb, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0x125
		net.(*TCPConn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0xc2083383f0, 0x0, 0x0)
--
I0623 02:18:50.532899     337 client.go:121] client 127.0.0.1:53246 connected
I0623 02:18:50.533373     337 client.go:121] client 127.0.0.1:60960 connected
I0623 02:18:50.533807     337 client.go:121] client 127.0.0.1:53416 connected
I0623 02:18:50.534239     337 client.go:121] client 127.0.0.1:38872 connected
I0623 02:18:50.534711     337 client.go:121] client 127.0.0.1:43122 connected
--- FAIL: TestComplexScenarios (9.33s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080118e0, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x63
		net.(*pollDesc).WaitRead(0xc2080118e0, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x51
		net.(*netFD).Read(0xc208011880, 0xc2083aa800, 0x400, 0x400, 0x0, 0x7f9b4ee1fd80, 0xc2084aa018)
			/usr/src/go/src/net/fd_unix.go:242 +0x4b3
		net.(*conn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0x4403fb, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0x125
		net.(*TCPConn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0xc2083383f0, 0x0, 0x0)
--
			github.com/cockroachdb/cockroach/rpc/_test/_testmain.go:104 +0x28d
=== RUN TestUpdatedAddr
W0623 02:18:50.645354     337 server.go:212] asked for port 1234, got 1235
W0623 02:18:50.645511     337 server.go:212] asked for port 1234, got 1235
W0623 02:18:50.645642     337 server.go:212] asked for port 1234, got 1235
--- FAIL: TestUpdatedAddr (0.10s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080118e0, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x63
		net.(*pollDesc).WaitRead(0xc2080118e0, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x51
		net.(*netFD).Read(0xc208011880, 0xc2083aa800, 0x400, 0x400, 0x0, 0x7f9b4ee1fd80, 0xc2084aa018)
			/usr/src/go/src/net/fd_unix.go:242 +0x4b3
		net.(*conn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0x4403fb, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0x125
		net.(*TCPConn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0xc2083383f0, 0x0, 0x0)
--
			/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x36
		github.com/cockroachdb/cockroach/rpc.TestMain(0xc20800a910)
			/go/src/github.com/cockroachdb/cockroach/rpc/main_test.go:56 +0x36
		main.main()
			github.com/cockroachdb/cockroach/rpc/_test/_testmain.go:104 +0x28d
FAIL
FAIL	github.com/cockroachdb/cockroach/rpc	14.481s
=== RUN TestAll
I0623 02:18:50.862287     392 rpc_test.go:49] Arith.Add(1, 2): 3
I0623 02:18:50.867862     392 rpc_test.go:49] Arith.Add(1, 2): 3
--- PASS: TestAll (0.01s)
PASS
ok  	github.com/cockroachdb/cockroach/rpc/codec	0.038s
?   	github.com/cockroachdb/cockroach/rpc/codec/message	[no test files]
?   	github.com/cockroachdb/cockroach/rpc/codec/wire	[no test files]
=== RUN TestGetCertificateUser
--- PASS: TestGetCertificateUser (0.00s)
=== RUN TestUnretryableError
I0623 02:18:40.893453     337 client.go:121] client 127.0.0.1:51115 connected
--- PASS: TestUnretryableError (0.32s)
=== RUN TestClientNotReady
E0623 02:18:41.207007     337 client.go:118] client 127.0.0.1:58314 failed to connect: client.go:303: client is closed
--- FAIL: TestClientNotReady (0.41s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080118e0, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x63
		net.(*pollDesc).WaitRead(0xc2080118e0, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x51
		net.(*netFD).Read(0xc208011880, 0xc2083aa800, 0x400, 0x400, 0x0, 0x7f9b4ee1fd80, 0xc2084aa018)
			/usr/src/go/src/net/fd_unix.go:242 +0x4b3
		net.(*conn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0x4403fb, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0x125
		net.(*TCPConn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0xc2083383f0, 0x0, 0x0)
--
I0623 02:18:50.532899     337 client.go:121] client 127.0.0.1:53246 connected
I0623 02:18:50.533373     337 client.go:121] client 127.0.0.1:60960 connected
I0623 02:18:50.533807     337 client.go:121] client 127.0.0.1:53416 connected
I0623 02:18:50.534239     337 client.go:121] client 127.0.0.1:38872 connected
I0623 02:18:50.534711     337 client.go:121] client 127.0.0.1:43122 connected
--- FAIL: TestComplexScenarios (9.33s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080118e0, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x63
		net.(*pollDesc).WaitRead(0xc2080118e0, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x51
		net.(*netFD).Read(0xc208011880, 0xc2083aa800, 0x400, 0x400, 0x0, 0x7f9b4ee1fd80, 0xc2084aa018)
			/usr/src/go/src/net/fd_unix.go:242 +0x4b3
		net.(*conn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0x4403fb, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0x125
		net.(*TCPConn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0xc2083383f0, 0x0, 0x0)
--
			github.com/cockroachdb/cockroach/rpc/_test/_testmain.go:104 +0x28d
=== RUN TestUpdatedAddr
W0623 02:18:50.645354     337 server.go:212] asked for port 1234, got 1235
W0623 02:18:50.645511     337 server.go:212] asked for port 1234, got 1235
W0623 02:18:50.645642     337 server.go:212] asked for port 1234, got 1235
--- FAIL: TestUpdatedAddr (0.10s)
	<autogenerated>:31: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080118e0, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x63
		net.(*pollDesc).WaitRead(0xc2080118e0, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x51
		net.(*netFD).Read(0xc208011880, 0xc2083aa800, 0x400, 0x400, 0x0, 0x7f9b4ee1fd80, 0xc2084aa018)
			/usr/src/go/src/net/fd_unix.go:242 +0x4b3
		net.(*conn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0x4403fb, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0x125
		net.(*TCPConn).Read(0xc208050160, 0xc2083aa800, 0x400, 0x400, 0xc2083383f0, 0x0, 0x0)
--
			/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x36
		github.com/cockroachdb/cockroach/rpc.TestMain(0xc20800a910)
			/go/src/github.com/cockroachdb/cockroach/rpc/main_test.go:56 +0x36
		main.main()
			github.com/cockroachdb/cockroach/rpc/_test/_testmain.go:104 +0x28d
FAIL
FAIL	github.com/cockroachdb/cockroach/rpc	14.481s
=== RUN TestAll
I0623 02:18:50.862287     392 rpc_test.go:49] Arith.Add(1, 2): 3
I0623 02:18:50.867862     392 rpc_test.go:49] Arith.Add(1, 2): 3
--- PASS: TestAll (0.01s)
PASS
ok  	github.com/cockroachdb/cockroach/rpc/codec	0.038s
?   	github.com/cockroachdb/cockroach/rpc/codec/message	[no test files]
?   	github.com/cockroachdb/cockroach/rpc/codec/wire	[no test files]
=== RUN TestGetCertificateUser
--- PASS: TestGetCertificateUser (0.00s)

```
Please assign, take a look and update the issue accordingly.

The full stacktrace shows that this is a request being read, with nothing cockroach-specific running any more (except for the codec). I suspect that this could happen even if everything has been closed (closing the client closes a channel, closing the codec closes its `baseConn`, which does nothing but close the underlying connection - I think that ends up here (but there were a lot of jumps)

```
// Close stops listening on the TCP address.
// Already Accepted connections are not closed.
func (l *TCPListener) Close() error {
	if l == nil || l.fd == nil || l.fd.ctl == nil {
		return syscall.EINVAL
	}
	if _, err := l.fd.ctl.WriteString("hangup"); err != nil {
		l.fd.ctl.Close()
		return &OpError{"close", l.fd.ctl.Name(), l.fd.laddr, err}
	}
	return l.fd.ctl.Close()
}
```

What seems weird is that TestComplexScenarios clocks in at `9.33s`: `--- FAIL: TestComplexScenarios (9.33s)`, which seems long.

Does anybody else have insights into this?

--- a/rpc/server.go
+++ b/rpc/server.go
@@ -154,7 +154,14 @@ func (s *Server) Serve(handler http.Handler) {
 
 			switch state {
 			case http.StateNew:
-				s.activeConns[conn] = struct{}{}
+				select {
+				case <-s.context.Stopper.ShouldStop():
+					// Shutting down? Just close the new connection immediately.
+					conn.Close()
+				default:
+					s.activeConns[conn] = struct{}{}
+				}
+
 			case http.StateHijacked, http.StateClosed:
 				delete(s.activeConns, conn)
 			}

